linker/editions.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linker
+
+import (
+	"fmt"
+	"sync"
+
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+var (
+	editionDefaults     map[descriptorpb.Edition]*descriptorpb.FeatureSet
+	editionDefaultsInit sync.Once
+)
+
+type hasFeatures interface {
+	GetFeatures() *descriptorpb.FeatureSet
+}
+
+var _ hasFeatures = (*descriptorpb.FileOptions)(nil)
+var _ hasFeatures = (*descriptorpb.MessageOptions)(nil)
+var _ hasFeatures = (*descriptorpb.FieldOptions)(nil)
+var _ hasFeatures = (*descriptorpb.OneofOptions)(nil)
+var _ hasFeatures = (*descriptorpb.ExtensionRangeOptions)(nil)
+var _ hasFeatures = (*descriptorpb.EnumOptions)(nil)
+var _ hasFeatures = (*descriptorpb.EnumValueOptions)(nil)
+var _ hasFeatures = (*descriptorpb.ServiceOptions)(nil)
+var _ hasFeatures = (*descriptorpb.MethodOptions)(nil)
+
+// resolveFeature resolves a feature for the given descriptor. It uses the function
+// to extract a value from a feature set. The function should return false as the
+// second argument if the first value it returns was not actually present in the
+// feature set.
+func resolveFeature[T any](d protoreflect.Descriptor, accessor func(*descriptorpb.FeatureSet) (T, bool)) T {
```

Comment:
Could the accessor be of type `func(*descriptorpb.FeatureSet) *T` and pre check for nil `FeatureSet` with the final defaults being a guaranteed non-nil? Would replace:
```
func(features *descriptorpb.FeatureSet) (descriptorpb.FeatureSet_FieldPresence, bool) {
	return features.GetFieldPresence(), features != nil && features.FieldPresence != nil
})
```
```
func(features *descriptorpb.FeatureSet) (*descriptorpb.FeatureSet_FieldPresence) {
	return features.FieldPresence
})
```

Reply:
If an explicit `default` were ever added to any feature, that wouldn't work -- we'd want to use the `Get*` accessor so it returns the default.

But I guess another formulation for this is to use protobuf reflection and pass in a `protoreflect.FieldDescriptor` instead of a lambda, and that would make the call sites simpler.

Reply:
Done